crates/sdk/src/websocket.rs

```diff
@@ -313,15 +355,42 @@ impl WsConnection {
                     }
 
                     Ok(Some(WebSocketMessage::Ping(payload))) => {
+                        log::trace!("received ping");
+                        idle = false;
                         record_metrics(payload.len());
-                    }
+                    },
+
+                    Ok(Some(WebSocketMessage::Pong(payload))) => {
+                        log::trace!("received pong");
+                        idle = false;
+                        want_pong = false;
+                        record_metrics(payload.len());
+                    },
 
                     Ok(Some(other)) => {
                         log::warn!("Unexpected WebSocket message {:?}", other);
+                        idle = false;
                         record_metrics(other.len());
                     },
                 },
 
+                _ = idle_timeout_interval.tick() => {
+                    if mem::replace(&mut idle, true) {
+                        if mem::take(&mut want_pong) {
+                            // Nothing received while we were waiting for a pong.
+                            log::warn!("Connection timed out");
+                            break;
+                        }
+
+                        log::trace!("sending client ping");
+                        Self::maybe_log_error(
+                            "Error sending ping",
+                            self.sock.send(WebSocketMessage::Ping(vec![])).await,
+                        );
+                        want_pong = true;
```

Comment:
Do we want to break the loop if we fail to send a ping?

Reply:
Good point. I think this applies to all send attempts, though, so I'm adding it to the outgoing branch too.

Reply:
Sounds good.